include/zephyr/fs/zms.h

```diff
@@ -0,0 +1,212 @@
+	const struct device *flash_device;
+	/** Flash memory parameters structure */
+	const struct flash_parameters *flash_parameters;
+#if CONFIG_ZMS_LOOKUP_CACHE
+	uint64_t lookup_cache[CONFIG_ZMS_LOOKUP_CACHE_SIZE];
+#endif
+};
+
+/**
+ * @}
+ */
+
+/**
+ * @brief Zephyr Memory Storage APIs
+ * @defgroup zms_high_level_api Zephyr Memory Storage APIs
+ * @ingroup zms
+ * @{
+ */
+
+/**
+ * @brief Mount a ZMS file system onto the device specified in @p fs.
+ *
+ * @param fs Pointer to file system
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int zms_mount(struct zms_fs *fs);
+
+/**
+ * @brief Clear the ZMS file system from device.
+ *
+ * @param fs Pointer to file system
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int zms_clear(struct zms_fs *fs);
+
+/**
+ * @brief Write an entry to the file system.
+ *
+ * @note  When @p len parameter is equal to @p 0 then entry is effectively removed (it is
+ * equivalent to calling of zms_delete). It is not possible to distinguish between a deleted
+ * entry and an entry with data of length 0.
+ *
+ * @param fs Pointer to file system
+ * @param id Id of the entry to be written
+ * @param data Pointer to the data to be written
+ * @param len Number of bytes to be written (maximum 64 KB)
+ *
+ * @return Number of bytes written. On success, it will be equal to the number of bytes requested
+ * to be written. When a rewrite of the same data already stored is attempted, nothing is written
+ * to flash, thus 0 is returned. On error, returns negative value of errno.h defined error codes.
+ */
+ssize_t zms_write(struct zms_fs *fs, uint32_t id, const void *data, size_t len);
+
+/**
+ * @brief Delete an entry from the file system
+ *
+ * @param fs Pointer to file system
+ * @param id Id of the entry to be deleted
+ * @retval 0 Success
+ * @retval -ERRNO errno code if error
+ */
+int zms_delete(struct zms_fs *fs, uint32_t id);
+
+/**
+ * @brief Read an entry from the file system.
+ *
+ * @param fs Pointer to file system
+ * @param id Id of the entry to be read
+ * @param data Pointer to data buffer
+ * @param len Number of bytes to be read (or size of the allocated read buffer)
+ *
+ * @return Number of bytes read. On success, it will be equal to the number of bytes requested
+ * to be read. When the return value is less than the number of bytes requested to read this
+ * indicates that ATE contain less data than requested. On error, returns negative value of
+ * errno.h defined error codes.
+ */
+ssize_t zms_read(struct zms_fs *fs, uint32_t id, void *data, size_t len);
+
+/**
+ * @brief Read a history entry from the file system.
+ *
+ * @param fs Pointer to file system
+ * @param id Id of the entry to be read
+ * @param data Pointer to data buffer
+ * @param len Number of bytes to be read
+ * @param cnt History counter: 0: latest entry, 1: one before latest ...
+ *
+ * @return Number of bytes read. On success, it will be equal to the number of bytes requested
```

Comment:
side note: do you know if this API has any real-life use case? I wonder if having this in a public API brings any value.

Reply:
What does a `history entry` even mean? This is not clear from the prototype comment.
If this doesn't necessarily have a use case (I don't know about that), I would suggest maybe keeping this function internal for now?